notifyme/notifyme.1
.TH NOTIFYME 1 "2010-03-10"
.SH NAME
notifyme - ldm NOTIFYME example client
.SH "UNIX SYNOPSIS"
\fBnotifyme \fR[\fB\-vxyz\fR] [\fB\-l \fIlogdest\fR] 
.BI [\-h " remote" ]
.BI [\-p " pattern" ]
.BI [\-f " feedtype" ]
.BI [\-o " offset" ]
.BI [\-t " timeout" ]
.BI [\-T " TotalTimeo" ]
.SH DESCRIPTION
This program receives notification messages from the 
Unidata Local Data Management system (ldm) and optionally
prints the notification information with timestamps.
It establishes a connection with the LDM server at \fIremote\fR and requests
notification of the arrival data whose identifier matches \fIpattern\fR.
As data arrives at the server,  notification is sent to this program.
.LP
This program is an example of how to use ldm4 notification.
.SH OPTIONS 
.TP 8
.BI \-f " feedtype"
Requests that notification of data that originates in a feed type of type 
\fIfeedtype\fP be sent along.
The default is `ANY'.
.TP
.BI \-h " remote"
Requests notification from the ldm server on \fIremote\fR. Default
is "localhost".
.TP
.BI \-l " logdest"
Log to \fIlogdest\fP. One of \fB''\fP (system logging daemon), \fB'-'\fP
(standard error stream), or file \fIlogdest\fP. Default is the standard error
stream if the process has a controlling terminal (i.e., the process isn't a
daemon); otherwise, either the LDM log file or the system logging daemon
(execute this program with just the option \fB'-?'\fP to determine which).
.TP
.BI \-o " offset"
Requests notification for data that arrived \fIoffset\fP seconds before NOW.
The default is 0.
.TP
.BI \-p " pattern"
Requests that notification of data be sent along that matches \fIpattern\fP.
The default is `\fB.*\fP'.
.TP
.BI \-T " TotalTimeo"
Give up attempting to make a connection after
\fITotalTimeo\fR seconds. Any partially built up client connnection
is destroyed and we start over.
.TP
.BI \-t " timeout"
Sets the RPC timeout on FEEDME requests to \fItimeout\fR seconds.
.TP
.B \-v
Verbose flag.  If present, the log will contain a line for 
each product that was received. Each line reports (1) the current system time, 
(2) the string 'notifyme', (3) the product size, (4) the time the product 
was injected into the IDD, (5) the feed type, (6) a sequence number, and 
(7) the product ID with the actual product time. (If you take the difference 
between the first and fourth fields, you can see how long it took the product 
to arrive at the host.)
.TP
.BI \-x
Requests that debugging information also be printed.
.TP
.B -y
Use logging timestamps with microsecond resolution. The default is
second resolution.
.TP
.B -z
Use logging timestamps formatted according to the ISO 8601 standard
(e.g.
.IR YYYYMMDD T hhmmss ).
The default is
.IR Mon\ day\ hh : mm : ss
.SH EXAMPLE
The command
.RS +4
.nf
notifyme -v
.fi
.RE
will let you watch the data come in to the local host.
.SH "SEE ALSO"
ldmd(1), ldmd.conf(5), feedme(1) 
.SH DIAGNOSTICS
Error messages and log messages are written to the
log file. 
.SH BUGS and LIMITATIONS
While this is the most important diagnostic tool in the LDM package, if
you leave off the -v flag, you won't see much.


CHANGE_LOG
6.13.0.0	2016-02-11T12:02:01-0700
    Logging:
        Re-wrote LDM logging because:
            1) It was the major impediment to correctly using the LDM;
            2) It was the major impediment to putting the LDM in a Docker
               container; and
            3) Some system logging daemons sometimes dropped LDM log messages.
        For backward-compatibility, the original logging component, "ulog",
        still exists and is unconditionally included in the LDM library,
        "libldm". Indeed, the original behavior of the LDM can be obtained by
        specifying the option "--with-ulog" to the configure(1) script. By
        default, however, the new logging component, named "slog" (for "simple
        logging"), will be built and used. While "slog" can be told to write to
        the system logging daemon, by default it will write directly to the LDM
        log file if the process doesn't have a controlling terminal (i.e., is a
        daemon) or to the standard error stream if the process does have a
        controlling terminal (i.e., is not a daemon). Like the original "ulog"
        component, "slog" also supports log file rotation via a
        crontab(1)-executed "ldmadmin newlog" command. See the installed HTML
        documentation "$HOME/share/doc/ldm/basics/logging-format.html" or the
        web page
        <http://www.unidata.ucar.edu/software/ldm/ldm-current/basics/logfile-format.html>
        for more information.

6.12.15.42	2016-01-12T10:40:24-0700
    Make test for C++11-compatible g++ more robust.

6.12.15.41	2016-01-12T09:28:01-0700
    Get "make distclean" to work on both Winters (with C++11 support) and Spock
    (without C++11 support).

6.12.15.40	2016-01-11T16:07:56-0700
    Removed mylog2ulog stuff from ulog makefile.
    
    Eliminated memory-leaks from components that read GEMPAK tables.

6.12.15.39	2015-11-16T16:04:41-0700
    mldm_sender_manager(3):
        Add "-y" and "-z" options to execvp(2) of mldm_sender(1).

    pqact.c:
        Added SIGHUP forwarding.

6.12.15.38	2015-11-03T13:07:13-0700
    ldmd.c:
        Corrected some Coverity-discovered lint.

6.12.15.37	2015-11-03T09:41:04-0700
    Increase resolution of printed product-creation timestamp from millisecond
    to microsecond.

6.12.15.36	2015-11-02T15:59:18-0700
    ldmadmin(1):
        Reduce overestimation of new queue capacities (i.e., data capacity,
        product capacity) and don't compute new capacities if the data is
        insufficient.
        
    logging:
        Added logging options for using microsecond resolution in timestamps and
        formatting them according to the ISO 8601 standard. Added respective
        options "-y" and "-z" to ldmd(1) and notifyme(1).

6.12.15.35	2015-10-28T15:29:59-0600
    Create mcast_test(1) for testing LDM-7 on GENI.

6.12.15.34	2015-10-26T16:41:04-0600
    PerProdNotifier:
        Fix setting of product-information by notify_of_bop().

6.12.15.33	2015-10-23T15:57:05-0600
    mldm_sender():
        Moved addition of product-index-to-product-signature entry to before the
        sending of the product because the receiving LDM-7 would sometimes
        request a data-product before the mapping was entered resulting in the
        product not being found.

6.12.15.32	2015-10-22T19:32:42-0600
    Improve receiver session-memory API.

6.12.15.31	2015-10-20T17:11:53-0600
    down7.c:
        Correct location of down7_incNumProds().

        Add mutex destruction to down7_free().

    mldm_sender:
        Add opening of product-queue in thread-safe mode to prevent corruption
        of the regions-in-use list by multiple thread access.

    pq.c:
        Add unlocking of mutex in pq_sequence() before calling `ifMatch()`
        to avoid deadlock in calling a foreign function.

        Add mutex destruction to freeing.

6.12.15.30	2015-10-19T16:23:06-0600
    noaaportIngester(1):
        Added "+000 001 008 010 Total precipitation" to GEMPAK tables.

    Multicast:
        Add feature to lock products until VCMTP sender releases them.

        Modify VCMTP port numbers to match WireShark plugin.

        Improve log messages.

    pq(3):
        Make pq_release() thread-safe.

6.12.15.29	2015-10-08T18:03:25-0600
    Add diagnostic print statements and modify up7-down7_test to try
    diagnosing product-queue corruption seen by Shawn chen.

6.12.15.28	2015-10-06T16:03:42-0600
    pq(3):
        Modified fb_get() to log fblk metrics when it fails to help diagnose
        multicast problem.

6.12.15.27	2015-10-06T12:01:45-0600
    pqutil(1):
        Removed minor lint revealed by Coverity Scan.

    pq(3):
        Use correct product size in pqe_insert() and pq_insertNoSig().

    mcast_lib(3):
        Adapt to new checks by pqe_insert().

6.12.15.26	2015-10-06T09:48:58-0600
    pqact(1):
        Correct the setting of the product-queue cursor based on the last
        successfully-processed data-product from the previous session.
        Before this, pqact(1) might miss a product due to a restart of the LDM.
        This was most apparent in processing NEXRAD-2 data-products. Thanks
        Ryan.

    Product-Queue (pq(3)):
        Refactored pqe_insert() for better error-handling and return status.

        Harmonized return code names by adding PQ_DUP and PQ_BIG and making
        PQUEUE_DUP and PQUEUE_BIG aliases.

6.12.15.25	2015-10-01T15:28:56-0600
    Remove ignoring of SIGCONT from mldm_sender_manager because it's not needed.

    Generalize TCP keep-alive in TcpSend.cpp.

6.12.15.24	2015-09-30T15:40:21-0600
    Reduce number of "make distcheck" to just one in CI-commit.sh.

6.12.15.23	2015-09-30T14:33:58-0600
    Replace "$@" with just $@ in deploy.sh.

6.12.15.22	2015-09-30T13:32:25-0600
    Replace ${configOpts} with "$@" in deploy.sh to accommodate argument with
    space.

6.12.15.21	2015-09-29T16:57:17-0600
    Corrected setting of configure(1) options in delivery/deploy.sh.

6.12.15.20	2015-09-29T15:03:45-0600

notifyme/notifyme.c
/*
 *   Copyright 1993, University Corporation for Atmospheric Research
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: notifyme.c,v 1.63.12.8 2008/04/15 16:34:07 steve Exp $ */

/* 
 * get notification
 */

#include <config.h>

#include "ldm.h"
#include "ldm5.h"
#include "globals.h"
#include "remote.h"
#include "ldmprint.h"
#include "atofeedt.h"
#include "log.h"
#include "inetutil.h"
#include "ldm5_clnt.h"
#include "RegularExpressions.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <sys/socket.h>
#include <errno.h>
#include <regex.h>

#ifdef NO_ATEXIT
#include "atexit.h"
#endif

#ifndef DEFAULT_REMOTE
#define DEFAULT_REMOTE "localhost"
#endif
#ifndef DEFAULT_TIMEO
#define DEFAULT_TIMEO  25
#endif
#ifndef DEFAULT_TOTALTIMEO
#define DEFAULT_TOTALTIMEO  (12*DEFAULT_TIMEO)
#endif
#ifndef DEFAULT_FEEDTYPE
#define DEFAULT_FEEDTYPE ANY
#endif
#ifndef DEFAULT_PATTERN
#define DEFAULT_PATTERN ".*"
#endif


static const char *remote = DEFAULT_REMOTE; /* hostname of data remote */
static ldm_replyt reply = { OK };


/*
 * Called at exit.
 * This callback routine registered by atexit().
 */
static void
cleanup(void)
{
        log_notice("exiting");

        /* TODO: sign off */

        (void)log_fini();
}


/*
 * Called upon receipt of signals.
 * This callback routine registered in set_sigactions() .
 */
static void
signal_handler(int sig)
{
#ifdef SVR3SIGNALS
        /* 
         * Some systems reset handler to SIG_DFL upon entry to handler.
         * In that case, we reregister our handler.
         */
        (void) signal(sig, signal_handler);
#endif
        switch(sig) {
        case SIGHUP :
                log_refresh();
                return;
        case SIGINT :
                /*FALLTHROUGH*/
        case SIGTERM :
                done = !0;
                return;
        case SIGUSR1 :
                return;
        case SIGUSR2 :
                log_roll_level();
                return;
        case SIGPIPE :
                return;
        }
}


static void
set_sigactions(void)
{
#ifdef HAVE_SIGACTION
        struct sigaction sigact;

        sigact.sa_handler = signal_handler;
        sigemptyset(&sigact.sa_mask);
        sigact.sa_flags = SA_RESTART;

        (void) sigaction(SIGHUP,  &sigact, NULL);

        sigact.sa_flags = 0;
        (void) sigaction(SIGINT,  &sigact, NULL);
        (void) sigaction(SIGTERM, &sigact, NULL);
        (void) sigaction(SIGUSR1, &sigact, NULL);
        (void) sigaction(SIGUSR2, &sigact, NULL);
        (void) sigaction(SIGPIPE, &sigact, NULL);
#else
        (void) signal(SIGHUP, signal_handler);
        (void) signal(SIGINT, signal_handler);
        (void) signal(SIGTERM, signal_handler);
        (void) signal(SIGUSR1, signal_handler);
        (void) signal(SIGUSR2, signal_handler);
        (void) signal(SIGPIPE, signal_handler);
#endif
}


static void
usage(
        char *av0  /*  id string */
)
{
        (void)fprintf(stderr,
"Usage: %s [options] \t\nOptions:\n", av0);
        (void)fprintf(stderr,
"\t-v             Verbose, report each notification\n");
        (void)fprintf(stderr,
"\t-x             Debug mode\n");
        (void)fprintf(stderr,
"\t-l dest        Log to `dest`. One of: \"\" (system logging daemon), \"-\"\n"
"\t               (standard error), or file `dest`. Default is \"%s\"\n",
                log_get_default_destination());
        (void)fprintf(stderr,
"\t-h remote      Have \"remote\" send us data (default \"%s\")\n",
                DEFAULT_REMOTE);
        (void)fprintf(stderr,
"\t-f feedtype    Interested in products from feed \"feedtype\" (default %s)\n",
        s_feedtypet(DEFAULT_FEEDTYPE));
        (void)fprintf(stderr,
"\t-p pattern     Interested in products matching \"pattern\" (default \"%s\")\n",
                DEFAULT_PATTERN);
        (void)fprintf(stderr,
"\t-o offset      Set the \"from\" time offset secs before now\n");
        (void)fprintf(stderr,
"\t-t timeout     Set RPC timeout to \"timeout\" seconds (default %d)\n",
                DEFAULT_TIMEO);
        (void)fprintf(stderr,
"\t-T TotalTimeo  Give up after this many secs (default %d)\n",
                DEFAULT_TOTALTIMEO);
        exit(1);
}


static prod_class clss;

/*
 * The RPC dispatch routine for this program.
 * Registered as a callback by svc_register() below.
 * Note that only NULLPROC and NOTIFICATION rpc procs are
 * handled by this program.
 */
static void
notifymeprog_5(struct svc_req *rqstp, SVCXPRT *transp)
{
        prod_info notice;

        switch (rqstp->rq_proc) {

        case NULLPROC:
                (void)svc_sendreply(transp, (xdrproc_t)xdr_void, (caddr_t)NULL);
                return;

        case NOTIFICATION:
                (void) memset((char*)&notice, 0, sizeof(notice));
                if (!svc_getargs(transp, (xdrproc_t)xdr_prod_info,
                    (caddr_t)&notice))
                {
                        svcerr_decode(transp);
                        return;
                }

                (void)exitIfDone(0);

                /*
                 * Update the request filter with the timestamp
                 * we just recieved.
                 * N.B.: There can still be duplicates after
                 * a reconnect.
                 */
                clss.from = notice.arrival;
                timestamp_incr(&clss.from);


                /* 
                 * your code here, example just logs it 
                 */
                log_info("%s", s_prod_info(NULL, 0, &notice,
                        log_is_enabled_debug));


                if(!svc_sendreply(transp, (xdrproc_t)xdr_ldm_replyt,
                    (caddr_t) &reply))
                {
                        svcerr_systemerr(transp);
                }

                (void)exitIfDone(0);

                if(!svc_freeargs(transp, xdr_prod_info, (caddr_t) &notice)) {
                        log_error("unable to free arguments");
                        exit(1);
                }
                /* no break */

        default:
                svcerr_noproc(transp);
                return;
        }
}


int main(int ac, char *av[])
{
        char *logfname = 0;
        unsigned timeo = DEFAULT_TIMEO; 
        unsigned interval = DEFAULT_TIMEO; 
        unsigned TotalTimeo = DEFAULT_TOTALTIMEO;
        prod_spec spec;
        int status;
        prod_class_t *clssp;
        unsigned        port = 0;
        unsigned        logOpts = LOG_CONS | LOG_PID;

        if(set_timestamp(&clss.from) != 0)
        {
                fprintf(stderr, "Couldn't set timestamp\n");
                exit(1);
        }
        clss.to = TS_ENDT;
        clss.psa.psa_len = 1;
        clss.psa.psa_val = &spec;
        spec.feedtype = DEFAULT_FEEDTYPE;
        spec.pattern = DEFAULT_PATTERN;

        { /* Begin getopt block */
        extern int optind;
        extern int opterr;
        extern char *optarg;
        int ch;
        int fterr;

        opterr = 1;

        while ((ch = getopt(ac, av, "vxl:f:o:t:h:P:p:T:")) != EOF)
                switch (ch) {
                case 'v':
                        if (!log_is_enabled_info)
                            log_set_level(LOG_LEVEL_INFO);
                        break;
                case 'x':
                        log_set_level(LOG_LEVEL_DEBUG);
                        break;
                case 'l':
                        logfname = optarg;
                        break;
                case 'h':
                        remote = optarg;
                        break;
                case 'P': {
                    char*       suffix = "";
                    long        p;

                    errno = 0;
                    p = strtol(optarg, &suffix, 0);

                    if (0 != errno || 0 != *suffix ||
                        0 >= p || 0xffff < p) {

                        (void)fprintf(stderr, "%s: invalid port %s\n",
                             av[0], optarg);
                        usage(av[0]);   
                    }
                    else {
                        port = p;
                    }

                    break;
                }
                case 'p':
                        spec.pattern = optarg;
                        /* compiled below */
                        break;
                case 'f':
                        fterr = strfeedtypet(optarg, &spec.feedtype);
                        if(fterr != FEEDTYPE_OK)
                        {
                                fprintf(stderr, "Bad feedtype \"%s\", %s\n",
                                        optarg, strfeederr(fterr));
                                usage(av[0]);   
                        }
                        break;
                case 'o':
                        clss.from.tv_sec -= atoi(optarg);
                        break;
                case 'T':
                        TotalTimeo = atoi(optarg);
                        if(TotalTimeo == 0)
                        {
                                fprintf(stderr, "%s: invalid TotalTimeo %s", av[0], optarg);
                                usage(av[0]);   
                        }
                        break;
                case 't':
                        timeo = (unsigned)atoi(optarg);
                        if(timeo == 0 || timeo > 32767)
                        {
                                fprintf(stderr, "%s: invalid timeout %s", av[0], optarg);
                                usage(av[0]);   
                        }
                        break;
                case '?':
                        usage(av[0]);
                        break;
                }

        if(ac - optind > 0)
                usage(av[0]);

        if (re_isPathological(spec.pattern))
        {
                fprintf(stderr, "Adjusting pathological regular-expression: "
                    "\"%s\"\n", spec.pattern);
                re_vetSpec(spec.pattern);
        }
        status = regcomp(&spec.rgx,
                spec.pattern,
                REG_EXTENDED|REG_NOSUB);
        if(status != 0)
        {
                fprintf(stderr, "Bad regular expression \"%s\"\n",
                        spec.pattern);
                usage(av[0]);
        }

        if(TotalTimeo < timeo)
        {
                fprintf(stderr, "TotalTimeo %u < timeo %u\n",
                         TotalTimeo, timeo);
                usage(av[0]);
        }

        } /* End getopt block */

        /*
         * initialize logger
         */
        (void)log_init(av[0]);
        log_notice("Starting Up: %s: %s",
                        remote,
                        s_prod_class(NULL, 0, &clss));

        /*
         * register exit handler
         */
        if(atexit(cleanup) != 0)
        {
                log_syserr("atexit");
                exit(1);
        }

        /*
         * set up signal handlers
         */
        set_sigactions();


        /*
         * Try forever.
         */
        while (exitIfDone(0))
        {
                clssp = &clss;
                status = forn5(NOTIFYME, remote, &clssp,
                                timeo, TotalTimeo, notifymeprog_5);

                (void)exitIfDone(0);

                switch(status) {
                        /* problems with remote, retry */       
                case ECONNABORTED:
                case ECONNRESET:
                case ETIMEDOUT:
                case ECONNREFUSED:
                        sleep(interval);
                        break;
                case 0:
                        /* assert(done); */
                        break;
                default:
                        /* some wierd error */
                        done = 1;
                        exit(1);
                }
        }

        exit(0);
        /*NOTREACHED*/
}
